Disallow modifier declarations and definitions in interfaces

cameel opened the pull request:
Fixes #11557.

cameel committed:
Disallow modifier declarations and definitions in interfaces

hrkrshnn commented:
Shouldn't this be breaking?

cameel commented:
We discussed it on the chat and @chriseth' opinion was that it should not.

chriseth commented:
Strictly it is breaking. @axic what do you think?

axic commented on test/libsolidity/analysis/FunctionCallGraph.cpp:
```diff
@@ -1114,7 +1114,6 @@ BOOST_AUTO_TEST_CASE(interfaces_and_abstract_contracts)
 	unique_ptr<CompilerStack> compilerStack = parseAndAnalyzeContracts(R"(
 		interface I {
 			event Ev(uint);
-			modifier m() virtual;
```
Damn, we even had a test for it?!

cameel replied:
Yeah, though when working on the `FunctionCallGraph` I just tried everything that compiled so it's not like it was by design.

axic commented:
It is a though question.  Since we never intended this to happen and rather started out with a very restricted feature set for interfaces (recall #3420 and many other related issues), it is easy to see that this is an oversight and a bug.

However given this likely existed for 2 years and that some forms worked, makes me wonder if anyone relies on it?

And if they do, what does that mean? Likely only that they are locked in to the last version if they insist on this feature.  If we want to be careful, we could ask others from the ecosystem (like the tooling channel).  If we are more confident that this is something nobody ever should have used, then we should go ahead in a non-breaking release and create a blog post to spread the word.

I'm leaning towards not making this a breaking change.

chriseth commented:
I'm also fine with calling it a bugfix.

chriseth commented:
The documentation does not mention modifiers, but it says that functions cannot be implemented, so I hope nobody thought that modifiers can be implemented... https://docs.soliditylang.org/en/v0.8.6/contracts.html#interfaces

Can you also update the documentation? It also seems that the bullet list somehow does not render properly.

axic commented:
> It also seems that the bullet list somehow does not render properly.

I mentioned this on matrix two weeks ago, none of the bullet points render anywhere.

cameel commented:
Rebased to resolve the changelog conflict and update docs.

> Can you also update the documentation? It also seems that the bullet list somehow does not render properly.

I updated the page to mention modifiers.

The bullet list issue seems to be a general issue with the docs now so I'm going to work on it in a separate PR. No need to hold back this one because of it.

chriseth commented on docs/contracts/interfaces.rst:
```diff
@@ -6,12 +6,14 @@
 **********
 
-Interfaces are similar to abstract contracts, but they cannot have any functions implemented. There are further restrictions:
+Interfaces are similar to abstract contracts, but they cannot have any functions or modifiers implemented.
```
If you say "modifiers implemented" here, it looks like it is OK to declare them.

cameel replied:
There's `They cannot declare modifiers` below. I wanted to emphasize that they cannot be defined either.

cameel replied:
I reworded it. How about now?

chriseth commented:
Now it's inconsistent with e.g. constructors, because there you only write "declare".

cameel commented:
ok. Changed to just "declare".

Marenz reviewed (changes requested):
LGTM, only needs a rebase and an update to the changelog to add them under the right version.

Marenz commented:
Rebased & fixed changelog entry position.
